core/storage/statistics/gae_models.py

```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
         cls.delete_multi(instances)
 
 
+class LearnerAnswerDetailsModel(base_models.BaseModel):
+    """Model for storing the answer details that a learner enters when they
+    are asked for it, per state one model instance is created. And currently
+    the model works for exploration and question.
+    """
+
+    # The maximum size in bytes that learner_answer_info_list can take,
+    # although the maximum size is 1000000 bytes, but 100000 bytes are
+    # left for other fields in the model.
+    _MAX_LEARNER_ANSWER_INFO_LIST_BYTE_SIZE = 900000
+
+    # The reference to the state for which model instance is created.
+    state_reference = ndb.StringProperty(required=True, indexed=True)
+    # The type of entity, that whether it is exploration, question, or any
+    # other entity if it is added in future.
+    entity_type = ndb.StringProperty(
+        required=True, indexed=True, choices=ENTITY_TYPE_CHOICES)
+    # The id of the interaction for which the answer details were received.
+    interaction_id = ndb.StringProperty(required=True, indexed=False)
+    # List of LearnerAnswerInfo dicts, which is defined in
+    # stats_domain.py, each dict corresponds to a single answer info of
+    # learner.
+    learner_answer_info_list = ndb.JsonProperty(repeated=True, indexed=False)
+    # The version of the learner_answer_info_list currently supported by Oppia.
+    # If the internal JSON structure of learner_answer_info_list changes,
+    # CURRENT_SCHEMA_VERSION in this class needs to be incremented.
+    schema_version = ndb.IntegerProperty(
+        indexed=True, default=(
+            feconf.CURRENT_LEARNER_ANSWERS_DETAILS_SCHEMA_VERSION))
+    # The total number of bytes needed to store all of the answers in the
+    # learner_answer_info_list. This value is found by summing the JSON
+    # sizes of all answer info dicts stored inside learner_answer_info_list.
+    accumulated_answer_info_json_size_bytes = ndb.IntegerProperty(
+        indexed=False, required=False, default=0)
+
+    @classmethod
+    def get_state_reference_for_exploration(cls, exp_id, state_name):
+        """Generate the state_reference for the state in an exploration.
+
+        Args:
+            exp_id: str. The id of the exploration.
+            state_name: str. The name of the state.
+
+        Returns:
+            str. The state_reference for a new instance of this class.
+        """
+        return '%s.%s' % (exp_id, state_name)
+
+    @classmethod
+    def get_state_reference_for_question(cls, question_id):
+        """Generate the state_reference for the state in the question.
+
+        Args:
+            question_id: str. The id of the question.
+
+        Returns:
+            str. The state_reference for a new instance of this class.
+        """
+        return question_id
+
+    @classmethod
+    def insert_learner_answer_info_for_exploration(
```

Comment:
No, you should not do this -- no logic in the storage layer. This should happen in the domain layer. Basically, the domain layer fetches an instance from the storage layer, modifies it, validates it, and puts it. The storage layer should just be a dumb interface that supports get/put/fetch.

(I know there's some logic in the other answers model but that's because that one deals with shards and supplementary storage models. You don't have such a constraint here.)

Reply:
Done